src/v/redpanda/application.cc

```diff
@@ -883,4 +923,71 @@ void application::admin_register_kafka_routes(ss::http_server& server) {
                   });
             });
       });
+
+    ss::httpd::partition_json::kafka_move_partition.set(
+      server._routes, [this](std::unique_ptr<ss::httpd::request> req) {
+          auto topic = model::topic(req->param["topic"]);
+
+          model::partition_id partition;
+          try {
+              partition = model::partition_id(
+                std::stoll(req->param["partition"]));
+          } catch (...) {
+              throw ss::httpd::bad_param_exception(fmt::format(
+                "Partition id must be an integer: {}",
+                req->param["partition"]));
+          }
+
+          if (partition() < 0) {
+              throw ss::httpd::bad_param_exception(
+                fmt::format("Invalid partition id {}", partition));
+          }
+
+          std::optional<std::vector<model::broker_shard>> replicas;
+          if (auto node = req->get_query_param("target"); !node.empty()) {
+              try {
+                  replicas = parse_target_broker_shards(node);
+              } catch (...) {
+                  throw ss::httpd::bad_param_exception(fmt::format(
+                    "Invalid target format {}: {}",
+                    node,
+                    std::current_exception()));
+              }
+          }
+
+          // this can be removed when we have more sophisticated machinary in
+          // redpanda itself for automatically selecting target node/shard.
+          if (!replicas || replicas->empty()) {
+              throw ss::httpd::bad_request_exception(
+                "Partition movement requires target replica set");
+          }
+
+          model::ntp ntp(model::kafka_namespace, topic, partition);
+
+          vlog(
+            _log.debug,
+            "Request to change ntp {} replica set to {}",
+            ntp,
+            replicas);
+
+          return controller->get_topics_frontend()
+            .local()
+            .move_partition_replicas(
+              ntp, *replicas, model::timeout_clock::now() + 5s)
+            .then([this, ntp, replicas](std::error_code err) {
+                vlog(
+                  _log.debug,
+                  "Result changing ntp {} replica set to {}: {}:{}",
+                  ntp,
+                  replicas,
+                  err,
+                  err.message());
+                if (err) {
```

Comment:
What if the partition movement was already initiated. I don’t get why that would be an exception. Should it simply say, a different http status but 200-OK. Partition movement already in progress or smith like that.

Reply:
This is the thinnest possible wrapper around the low-level API and this is reporting whatever error the controller is giving us. You're right that some of those errors will probably be convertible to 200 or a retry.

Reply:
Sounds good. I’m probably missing context. Just seems like we should know today what would be a retry-able error to me. But I will take your guidance here.

Reply:
You're not missing context, we probably do want to do that. But IMO it isn't certain until we get some experience with the api.

Reply:
Makes sense